index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndex.java

```diff
@@ -0,0 +1,308 @@
+            client.storeDocument(name(), workspace, nodeKey, doc);
+            client.refresh(name());
+        } catch (Exception e) {
+            throw new EsIndexException(e);
+        }
+    }
+
+    /**
+     * Searches indexed node's properties by node key.
+     * 
+     * @param nodeKey node key being indexed.
+     * @return list of stored properties as json document.
+     * @throws IOException 
+     */
+    private EsDocument find(String nodeKey) throws IOException {
+        return client.getDocument(name(), workspace, nodeKey);
+    }
+
+    /**
+     * Searches indexed node's properties by node key or creates new empty list.
+     * 
+     * @param nodeKey node key being indexed.
+     * @return list of stored properties as json document or empty document 
+     * if not found.
+     * @throws IOException 
+     */
+    private EsDocument findOrCreateDoc(String nodeKey) throws IOException {
+        EsDocument doc = client.getDocument(name(), workspace, nodeKey);
+        return doc != null ? doc : new EsDocument();
+    }
+
+    /**
+     * Appends specified value for the given column and related pseudo columns
+     * into list of properties.
+     * 
+     * @param doc list of properties in json format
+     * @param column colum definition
+     * @param value column's value.
+     */
+    private void putValue(EsDocument doc, EsIndexColumn column, Object value) {
+        Object columnValue = column.columnValue(value);
+        String stringValue = column.stringValue(value);
+        doc.put(column.getName(), columnValue);
+        doc.put(column.getLowerCaseFieldName(), stringValue.toLowerCase());
+        doc.put(column.getUpperCaseFieldName(), stringValue.toUpperCase());
+        doc.put(column.getLengthFieldName(), stringValue.length());
+    }
+
+    /**
+     * Appends specified values for the given column and related pseudo columns
+     * into list of properties.
+     * 
+     * @param doc list of properties in json format
+     * @param column colum definition
+     * @param value column's value.
+     */
+    private void putValues(EsDocument doc, EsIndexColumn column, Object[] value) {
+        Object[] columnValue = column.columnValues(value);
+        int[] ln = new int[columnValue.length];
+        String[] lc = new String[columnValue.length];
+        String[] uc = new String[columnValue.length];
+
+        for (int i = 0; i < columnValue.length; i++) {
+            String stringValue = column.stringValue(columnValue[i]);
+            lc[i] = stringValue.toLowerCase();
+            uc[i] = stringValue.toUpperCase();
+            ln[i] = stringValue.length();
+        }
+
+        doc.put(column.getName(), columnValue);
+        doc.put(column.getLowerCaseFieldName(), lc);
+        doc.put(column.getUpperCaseFieldName(), uc);
+        doc.put(column.getLengthFieldName(), ln);
+    }
+
+    @Override
+    public void commit() {
+        try {
+            client.flush(name());
```

Comment:
Again, perhaps I don't understand the Java API, but IIUC `client.flush(...)` is explicitly telling ES to flush all memory associated with an index to storage, and that ES tries to do this automatically. If we have to do this, then `commit()` is definitely the place to do it. But is it really required? Do we want to force a flush to get the durability of the write to disk?